File: 1419.minimum-number-of-frogs-croaking.cs
/*
 * @lc app=leetcode id=1419 lang=csharp
 *
 * [1419] Minimum Number of Frogs Croaking
 */

// @lc code=start
public class Solution {
    public int MinNumberOfFrogs(string croakOfFrogs) {
        Dictionary<Char,int> occur = new Dictionary<Char,int>();
        Dictionary<Char,int> frogs = new Dictionary<Char,int>();
        Dictionary<Char,Char> prevChar = new Dictionary<Char,Char>();
        int result = 0;

        occur.Add('c',0);
        occur.Add('r',0);
        occur.Add('o',0);
        occur.Add('a',0);
        occur.Add('k',0);
        frogs.Add('c',0);
        frogs.Add('r',0);
        frogs.Add('o',0);
        frogs.Add('a',0);
        frogs.Add('k',0);
        prevChar.Add('c','k');
        prevChar.Add('r','c');
        prevChar.Add('o','r');
        prevChar.Add('a','o');
        prevChar.Add('k','a');

        foreach (Char c in croakOfFrogs)
        {
            occur[c] += 1;
            if (occur['c'] < occur['r'] || occur['r'] < occur['o'] || occur['o'] < occur['a'] || occur['a'] < occur['k'])
            {
                return -1;
            }
        }
        if (occur['c'] != occur['r'] || occur['c'] != occur['o'] || occur['c'] != occur['a'] || occur['c'] != occur['k'])
            {
                return -1;
            }

        foreach (Char c in croakOfFrogs)
        {
            if (frogs[prevChar[c]] > 0)
            {
                frogs[prevChar[c]] -= 1;
            }
            else
            {
                result += 1;
            }
            frogs[c] += 1;
        }
        return result;
    }
}
// @lc code=end

File: 1419.minimum-number-of-frogs-croaking.py
#
# @lc app=leetcode id=1419 lang=python3
#
# [1419] Minimum Number of Frogs Croaking
#

# @lc code=start
class Solution:
    # First go through the string to ensure croak appears equal number of times
    # Also at any time, the occurrences of croak should be nonincreasing
    # Then we can simulate the process to match each character
    def minNumberOfFrogs(self, croakOfFrogs: str) -> int:
        occur = {'c': 0, 'r': 0, 'o': 0, 'a': 0, 'k': 0}
        for c in croakOfFrogs:
            occur[c] += 1
            if occur['c'] < occur['r'] or occur['r'] < occur['o'] or occur['o'] < occur['a'] or occur['a'] < occur['k']:
                return -1
        if occur['c'] != occur['r'] or occur['c'] != occur['o'] or occur['c'] != occur['a'] or occur['c'] != occur['k']:
            return -1
        
        frogs = {'c': 0, 'r': 0, 'o': 0, 'a': 0, 'k': 0}
        prev_char = {'c': 'k', 'r': 'c', 'o': 'r', 'a': 'o', 'k': 'a'}
        result = 0
        for c in croakOfFrogs:
            if frogs[prev_char[c]] > 0:
                frogs[prev_char[c]] -= 1
            else:
                result += 1
            frogs[c] += 1
        return result
# @lc code=end

